src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewFeatures/CookieTempDataProvider.cs

```diff
@@ -0,0 +1,86 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using Microsoft.AspNetCore.Cookies;
+using Microsoft.AspNetCore.DataProtection;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Mvc.ViewFeatures.Internal;
+using Microsoft.Extensions.Options;
+
+namespace Microsoft.AspNetCore.Mvc.ViewFeatures
+{
+    /// <summary>
+    /// Provides data from cookie to the current <see cref="ITempDataDictionary"/> object.
+    /// </summary>
+    public class CookieTempDataProvider : ITempDataProvider
+    {
+        public static readonly string CookieName = ".AspNetCore.Mvc.ViewFeatures.CookieTempDataProvider";
+        private static readonly string Purpose = "Microsoft.AspNetCore.Mvc.ViewFeatures.CookieTempDataProviderToken.v1";
+        private const byte TokenVersion = 0x01;
+        private readonly IDataProtector _dataProtector;
+        private TempDataSerializer _tempDataSerializer;
+        private readonly ChunkingCookieManager _chunkingCookieManager;
+        private readonly IOptions<CookieTempDataProviderOptions> _options;
+
+        public CookieTempDataProvider(IDataProtectionProvider dataProtectionProvider, IOptions<CookieTempDataProviderOptions> options)
+        {
+            _dataProtector = dataProtectionProvider.CreateProtector(Purpose);
+            _tempDataSerializer = new TempDataSerializer();
+            _chunkingCookieManager = new ChunkingCookieManager();
+            _options = options;
+        }
+
+        public IDictionary<string, object> LoadTempData(HttpContext context)
+        {
+            if (context == null)
+            {
+                throw new ArgumentNullException(nameof(context));
+            }
+
+            IDictionary<string, object> values = new Dictionary<string, object>(StringComparer.OrdinalIgnoreCase);
+            if (context.Request.Cookies.ContainsKey(CookieName))
+            {
+                var base64EncodedValue = _chunkingCookieManager.GetRequestCookie(context, CookieName);
```

Comment:
I'm sad about this, but I don't really see a better way right now. IMO this is should go in as-is and we can do perf as needed. None of this is public API